### ISSUE TITLE
Change detection (delta)

### ISSUE THREAD

**Issue by AnaBelgun** [TerriaJS/terriajs#3274]
Delta functionality (requested by MDBA/GA)
•	To develop a functionality which allows users to visualise changes for an area based on a calculation performed on (same?) layer at 2 different points in time
•	First stage: Sentinel to be used; later to extend to Landsat if possible
•	Roughly the user interaction:
1.	Select the near real time layer (extended to all Sentinel 2 a). For MVP use a calendar picker to select two dates (precursor to this would be zoom to a particular location, query the data for ‘data available for dates’ so that the user is presented with calendar pick options for which data is available.  Long term objective would be to integrate with hydrograph tool so that user can select ‘before and after’ in the context of the hydrograph rather than using a calendar picker
2.	Select the dates
3.	Turn on functionality (yes)
4.	Select (the) separate layer which shows the difference (using colour and the actual values) (yes)
5.	Export the difference as a full resolution (not just screen resolution) GeoTIFF to be used in own GIS and other systems (yes)

Some Q&A:
-	Does Terria need to perform a calculation per se (run an algorithm) (yes, this would be image array for date 1 – image array for date 2 calculation.  To add to the complexity here, the image differencing would be performed on data that need to be calculated on the fly.  For example the web services provide access to NIR and red band surface reflectance measurements.  So the client would need to a. calculate NDVI (NIR-RED/NIR+RED) for both dates, and then perform an image difference between the two dates (NDVI date 1 – NDVI date 2)
-	Is the difference something like: I see for this location an area coloured in blue on 1 Dec and for the same location, on 1 Jan, same area is coloured in red; a new layer will show that for such an area, there are changes in colours? - this is what the user is expecting

Example: 
NDMI 25/4/19
![NDMI 25-04-19](https://user-images.githubusercontent.com/18452161/57827597-429b3a80-77eb-11e9-9f4e-de632672ff42.png)

 
NDMI 05/05/2019
![NDMI 05-05-19](https://user-images.githubusercontent.com/18452161/57827603-46c75800-77eb-11e9-8283-aac14b5f7d1d.png)

 
NDMI Difference 05/05/2019 minus 25/04/2019
 
NDMI difference > 0.15 as red grid cells converted into 'uncleaned' polygons. Centre pivot polygons 'of interest' selected (light blue)
![NDMI difference uncleaned polygons](https://user-images.githubusercontent.com/18452161/57827710-bb9a9200-77eb-11e9-8317-abe67fc26132.png)

NDMI difference >0.15 (05/05/19 minus 25/04/19) final 'cleaned' irrigated 'areas of interest' polygons in red over aerial photo. Polygons emailed to field operations staff.
![NDMI difference cleaned polygons](https://user-images.githubusercontent.com/18452161/57827713-c2290980-77eb-11e9-9e44-a63559280eed.png)

**Comment by AnaBelgun** [comment]
need discussion if can be done in the app and if yes, how complex it is / how long will take

**Comment by AnaBelgun** [comment]
asked for more details about the difference calculation; pending reply

**Comment by na9da** [comment]
_This an evolving spec, changing as we get more details._

The delta tool will allow the user to visualize differences in imagery layers at two different points in time.

## Scope

1) The actual image diffing can happen on the WCS server. Terria will provide UI to trigger the diff with appropriate parameters

2) Polygon selection from raster data being a very hard problem, is not a part of the scope.

## Possible UX flow

1) Zoom to a region & add the near real time layer
2) Pick a date from the date picker (optionally enable 'Filter by location' to show just the dates for which imagery is available)
3) Click on 'Compare with another date` button (or simply Δ button).
4) This reveals a second date picker, pick a date and click 'Compare' button.
5) This creates a new catalog item, showing the delta image
8) Click Export button on the Catalog item to export as GeoTIFF.

## Implementation notes for the UX flow

3) Catalog definition must define a field `supportsDelta` to enable the functionality.



.

**Comment by AnaBelgun** [comment]
@philipgrimmett ping

**Comment by AnaBelgun** [comment]
The only additional requirement to keep in mind is the ability to pan across the landscape (looking at imagery along a river reach in detail).

**Comment by AnaBelgun** [comment]
from MDBA (6/06/2019): Thanks All,
I think this looks good too, we need to zoom in as well as pan of course, but presumably this is standard.
Would the tool calculate the statistics of the residual image? >5 sigma outliers = significant difference for example.
I think the choice of colour ramp would be important too, we might need some options to choose from.

**Comment by AnaBelgun** [comment]
@na9da @kring would you please approx effort for this one? by email pls

**Comment by na9da** [comment]
> from MDBA (6/06/2019): Thanks All,
> I think this looks good too, we need to zoom in as well as pan of course, but presumably this is standard.
> Would the tool calculate the statistics of the residual image? >5 sigma outliers = significant difference for example.
> I think the choice of colour ramp would be important too, we might need some options to choose from.

**Comment by AnaBelgun** [comment]
@AnaBelgun  to communicate estimation

**Comment by philipgrimmett** [comment]
My working theory:
Generating multiple fixed-location delta maps will be faster and less confusing overall than updating a viewport 'delta filter' every time you zoom/pan the map.

Invision concept for "box selection" workflow:
https://invis.io/CASND0RE5YT#/370254676_Delta_1

While in edit mode, the box selection should be scaled relative to the viewport ie. if you zoom in/out the box stays the same size. Once the delta map has been generated it should remain fixed to the globe and scale relative to the base map. (needs discussion)

**Pros**
1. You can auto-filter selected imagery by location and show only available dates
2. Conceptually it's a little easier to understand (eg. panning without a box introduces the problem of satellite paths not aligning with the selected dates, )
3. If the user wants to export to GeoTIFF they will probably have to select a region anyway

**Cons**
1. Suggestions?

**Options for updating/re-running the delta map function**
- On zoom, display help flag near delta map workbench item highlighting option to manually update
- On zoom, automatically re-run function after short delay if 1. delta map is within the current viewport and 2. tiles have changed (not sure if this is even sensible/possible @kring @na9da?)

**Comment by AnaBelgun** [comment]
pending DEA - MDBA decision on pursuing this;

**Comment by AnaBelgun** [comment]
@na9da @philipgrimmett 

Here are the latest comments on this: 
1.	User selects the data layer, navigates to desired area on the map at required zoom level and uses the (newly developed? part of the) interface to select two dates for which the delta image is required. Terria prepares the request in the same way as the original WMS GetMap request except that the time parameter includes two selected dates separated by comma before sending it to GSKY server.
 
2.	GSKY receives WMS GetMap requests in the following format:
http://server-ip/ows?time=2019-07-19T00%3A00%3A00.000Z,2019-07-21T00%3A00%3A00.000Z&srs=EPSG%3A3857&transparent=true&format=image%2Fpng&exceptions=application%2Fvnd.ogc.se_xml&styles=ndvi&tiled=true&feature_count=101&service=WMS&version=1.1.1&request=GetMap&layers=delta_service&bbox=15028131.257091936%2C-4383204.9499851465%2C15654303.392804097%2C-3757032.814272985&width=256&height=256
 
3.	GSKY sends back a PNG image same as any other WMS GetMap request.
4.	Terria refreshes user’s entire view port by replacing currently displayed tiles with the returned delta tiles
 
Please note:
a.	The WMS GetMap request for delta service is exactly the same as the original WMS GetMap request except that the time parameter now contains two dates separated by a comma.
b.	WMS getFeatureInfo returns the delta pixel value as the difference between the two dates.
c.	WMS GetCapabilities works in the same way as the original one.
d.	WMS DescribeLayer works in the same way as the original one.
e.	WMS GetLegendGraphic works in the same way as the original one.

**Comment by AnaBelgun** [comment]
'I’m happy for the MVP to be ‘a users zooms to a location of  interest, selects two dates (for which imagery is available) and GSKY returns an image delta for those two dates for Terria to refresh on the area on screen (view port)’.
 
In the longer run it’d would be ideal if you could pan along and the image delta would continue to be generated.
For example this location to start with, and then be able to pan down along the Darling River.
https://nationalmap.gov.au/#share=s-pcSRTXk0eSPfcjv23Q3fiemFrBK

**Comment by AnaBelgun** [comment]
"Once the delta service is up, this should be possible. For GSKY (and infact Terria), it is going to be similar to a typical GetMap request. When the user zooms in on a normal layer, Terria keeps generating more GetMap requests for GSKY at required zoom level and coordinates. Same way, for delta, when a user zooms in, Terria will send more GetMap requests to GSKY. The only difference in the requests will be that for a normal GetMap, one date is included but for delta, the request has two dates included. Hence, GSKY will keep responding with normal or delta image tiles accordingly."

**Comment by na9da** [comment]
To me it sounds like a regular WMS layer.

The only other concern is how to select the dates for which imagery is available. We should probably do the simplest thing first, which is making the user manually pick a point to filter the available dates. Once we get there, we can iteratively think of a better flow.

**Comment by philipgrimmett** [comment]
@nanda a few notes based on your UX flow build:

### Date formats in generated delta workbench item
For an Australian audience I think we should default to DD/MM/YYYY (currently YYYY/MM/DD in generated delta workbench items)

### Date selection
Users should only be able to open one dropdown to at a time, but I was able to open both when selecting primary/secondary image dates (see Chrome screenshot)

![Screen Shot 2019-11-04 at 2 24 58 pm](https://user-images.githubusercontent.com/5292858/68099231-6119fd00-ff15-11e9-985c-2dc3e53163fa.png)

Also, if possible we should set the following defaults for date selection:

**Primary image**
[date immediately preceding latest imagery]

**Secondary image**
[latest imagery]

This is a subtle nudge towards generating semi-meaningful imagery in cases where users just click though quickly without selecting unique dates (the primary and secondary fields currently default to the same date)

### Workflow title
When the change detection workflow is triggered, the following title should appear:
>Change detection: [source dataset title]

![Screen Shot 2019-11-04 at 3 24 03 pm](https://user-images.githubusercontent.com/5292858/68099588-62e4c000-ff17-11e9-9730-675eb26bf45d.png)

### Location selection marker
We'll have to display a temporary marker at the selected location to better communicate how the filter works (intersection with satellite imagery).

See https://invis.io/UBUCV5J2QDX#/388452859_Delta_2-1)

### lat/lon format
Probably just a bug, but the lat/lon values are currently prepended by  '$'

e.g. $-3834718.2136670854, $4828265.972742337

### Resolving availability at location
The 'resolving availability at location' alert remains on screen even after available dates are shown in the data panel. This might be a bug, but it raises an interesting point: what happens when users hastily click on the map, then decide they want to select a different location? The current workflow only allows users to proceed with their first selection.  It might be better if we allow users to select a new location at any point prior to clicking 'generate difference map'. This would involve reloading satellite imagery and updating available dates, but could help in situations where the filter returns a false positive (see screenshot)

![Screen Shot 2019-11-04 at 3 52 07 pm](https://user-images.githubusercontent.com/5292858/68100277-4f3b5880-ff1b-11e9-8819-053858211c0f.png)

### Feature info
When users click on generated delta imagery we currently display the lat/lon and primary date ('time') used to generate that imagery, which could be a little misleading given that delta imagery is a composite of two dates. Probably best if we display the following in feature info instead:
>Lat | [value]
>Lon | [value]
>primary image | [value]
>comparison image | [value]

![Screen Shot 2019-11-04 at 4 12 24 pm](https://user-images.githubusercontent.com/5292858/68101060-d38fda80-ff1f-11e9-817a-cc893bd0dd3c.png)

**Comment by na9da** [comment]
@philipgrimmett 

> Also, if possible we should set the following defaults for date selection:
> Primary image
> [date immediately preceding latest imagery]

We can also copy the date from the location filter/timeline to use as the primary date. Scenario: Suppose the user carefully zooms into a region and time and then decides to do a delta, he shouldn't have to choose the starting time again.

**Comment by na9da** [comment]
> Feature info
> When users click on generated delta imagery we currently display the lat/lon and primary date ('time') used to generate that imagery, which could be a little misleading given that delta imagery is a composite of two dates. Probably best if we display the following in feature info instead:

The time in the feature info panel is something we receive from the server. The real API we will use will ideally return the compared dates.

**Comment by philipgrimmett** [comment]
> We can also copy the date from the location filter/timeline to use as the primary date. Scenario: Suppose the user carefully zooms into a region and time and then decides to do a delta, he shouldn't have to choose the starting time again.

Hmmm, interesting! Is it reasonable for us to assume a user's current location filter represents their location of interest for delta imagery? I'd wager yes, but we'll have to make sure the pre-filled location is easy to update/override.

>The time in the feature info panel is something we receive from the server. The real API we will use will ideally return the compared dates.

We can template what appears in feature info though, yep?

**Comment by na9da** [comment]
> but we'll have to make sure the pre-filled location is easy to update/override.

I was suggesting copying just the date.

> We can template what appears in feature info though, yep?

Yes we can.

**Comment by philipgrimmett** [comment]
>I was suggesting copying just the date.

If we pre-fill the primary date from a pre-existing location filter, that's effectively the same as pre-filling the location, no?

**Comment by na9da** [comment]
Yes you are right, they are effectively the same. I'll leave that to the next round of fine tuning maybe :)

**Comment by na9da** [comment]
@philipgrimmett 

I have made the following changes:
- [x] Changed date format in the generated delta item to `dd-mm-yyyy`
- [x] Show only 1 date selector at a time
- [x] Dates in the date pickers will default to d-1 & d for primary and secondary dates
- [x] Changed tool title to show the catalog item name
- [x] Show a location selection marker
- [x] Properly format the coordinates
- [x] Have redone the user interaction code. User can now change his selection anytime.

Not done
- [ ] Formatting the feature info panel

Known issue:
* I am duplicating the original catalog item to use within the tool. Since the imagery takes a while to load, if the user selects a point on the map before the imagery has loaded, it will show `Error when trying to resolve imagery at location!` and then user has to retry. Ideally, we should allow the user to pick a point only after the layer has completely loaded. Have to check if that is possible.

**Comment by na9da** [comment]
Feedback from Leo to incorporate in next iteration:

> So the typical pattern of use for this capability would be:
> 1) Zoom to location
> 2) Select 2 cloud free images (typically from true colour or false colour image) - picking two cloud free images could be done via a window splitter (to pick the two dates)
> 3) Calculate index delta (for selected images)

The key point here is when selecting the dates, the user should be able to "browse through" and pick an image that is cloud free or has some other characteristic that he is looking for. This might mean, enabling the splitter when the tool is active and showing imagery for the selected dates on left and right.

...and 

>  you need to ensure that both before and after images are cloud free before you calculate the delta (currently need to resort to scribbling dates on a page prior to selecting the imagery).  Also you typically identify cloud free dates using true or false colour and having to remember to switch to NDVI prior to pushing the button is a pain.

Finding 2 cloud free dates and running delta on it is currently very tedious because you need to use true color style to choose the images, then select a style to run the delta on, then launch the tool, then select dates once again. We have to simplify that workflow.

**Comment by soyarsauce** [comment]
>Finding 2 cloud free dates and running delta on it is currently very tedious because you need to use true color style to choose the images, then select a style to run the delta on, then launch the tool, then select dates once again. We have to simplify that workflow.

This will definitely need a bit of love!

**Comment by rowanwins** [comment]
> Select 2 cloud free images (typically from true colour or false colour image) - picking two cloud free images could be done via a window splitter (to pick the two dates)

Currently there is nothing in the WMS that allows us to filter by cloud cover. 
I've chatted to DEA a few times about having some sort of metadata API but I don't think it's high on their roadmap last time I heard.

**Comment by soyarsauce** [comment]
>in the meantime, for delta, leave splitter as it is, re-wire delta to use a hybrid solution of deltav2 and delta v2.5(20200330 Sprint planning vic-demo)

**Comment by AnaBelgun** [comment]
Once Delta V3 is implemented we need to close v2 and v3 epics and move the remaining tickets into this generic Delta epic.

**Comment by soyarsauce** [comment]
Superseded by https://github.com/TerriaJS/terriajs/issues/4390
